src/main/java/org/omnifaces/util/JNDIObjectLocator.java

```diff
@@ -0,0 +1,266 @@
+ */
+
+package org.omnifaces.util;
+
+import java.io.Serializable;
+import java.util.Hashtable;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Pattern;
+import javax.ejb.Remote;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import lombok.AccessLevel;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.Lombok;
+import lombok.NonNull;
+import lombok.Singular;
+import lombok.SneakyThrows;
+import static org.omnifaces.util.JNDI.ENV_ENTRY_PREFIX;
+
+/**
+ * JNDIObjectLocator is used to centralize JNDI lookups. It minimizes the overhead of JNDI lookups by caching the
+ * objects it looks up.
+ * <p>
+ * Implements the ServiceLocator design pattern
+ *
+ * <pre>
+ *
+ * Major features are:
+ * - thread-safe
+ * - immutable
+ * - serializable
+ * - selectively disables the cache if objects are remote
+ *
+ * Examples:
+ *
+ * {@code locator = JNDIObjectLocator.builder().build();
+ * MyEJB myEJB1 = locator.getObject(MyEJB.class);
+ * MyEJB myEJB2 = locator.getObject("java:module/MyEJB");
+ * }
+ * </pre>
+ *
+ * <a href="https://github.com/flowlogix/flowlogix/blob/master/jakarta-ee/jee-examples/src/main/java/com/flowlogix/examples/JndiExample.java"
+ * target="_blank">Example Code (GitHub)</a>
+ *
+ * @author Lenny Primak
+ */
+@Builder(toBuilder = true)
+public class JNDIObjectLocator implements Serializable {
+    /**
+     * naming convention suffix for remote beans
+     */
+    public static final String REMOTE = "REMOTE";
+    /**
+     * naming convention suffix for local beans
+     */
+    public static final String LOCAL = "LOCAL";
+    static final String PORTABLE_NAME_PREFIX = "java:module";
+    /**
+     * pattern matcher for stripping local or remote suffix from beans
+     */
+    @SuppressWarnings("checkstyle:ConstantName")
+    public static final Pattern StripInterfaceSuffixPattern = Pattern.compile(LOCAL + "|" + REMOTE, Pattern.CASE_INSENSITIVE);
+
+    private static final long serialVersionUID = 2L;
+
+    /**
+     * to be passed into InitialContext()
+     */
+    private final @Singular("environment") Map<String, String> environment;
+    /**
+     * Used in construction of portable JNDI names
+     * usually java:module (default)
+     */
+    private final @Builder.Default String portableNamePrefix = PORTABLE_NAME_PREFIX;
+    /**
+     * whether to disable cache. Default is false
+     */
+    private final boolean noCaching;
+    /**
+     * whether to cache remote EJBs, usually false
+     */
+    private final boolean cacheRemote;
+
+    @Getter(AccessLevel.PACKAGE)
+    private final transient Map<String, Object> jndiObjectCache = new ConcurrentHashMap<>();
+    private final transient Lazy<InitialContext> initialContext = new Lazy<>(this::createInitialContext);
+    private final transient Lock initialContextLock = new ReentrantLock();
```

Comment:
I can't make sense of `initialContextLock`, can you explain it?

Reply:
According to this: https://stackoverflow.com/questions/13317848/is-javax-naming-initialcontext-threadsafe
lookups `InitialContext` are not thread-safe and need to be locked.
This is why all lookup calls with this possibly-shared, cached `InitialContext` are protected by this lock so they are not concurrently called.